File: src/commons/hub_state.py
import json
import time
from typing import Dict, Any, List
from logging import getLogger

from commons import constants as c

log = getLogger(__name__)

PERSISTED_STATE_FILE = "./persisted_state.json"
TEST_PERSISTED_STATE_INPUT_FILE = "./tests/persisted_state.json"
TEST_PERSISTED_STATE_OUTPUT_FILE = "./tests/persisted_state_out.json"


state: Dict[str, Any] = {
    # provided by central_hub/
    "hub_stats": {"state_updates_recv": 0},
    # last requested angles
    "set_angles": [],
    # actual angles last reported by servos
    "current_angles": [],
    # which subsystems are online and have indentified themselves
    "subsystem_stats": {},
    # array of json file names provided to central hub
    # by arms_config_provider
    "arm_config_files": [],
    # initially by arms_config_provider to last selected, can be changed
    # to any of the arm_configs
    "arm_config_selected": "",
    # currently selected arm config provided by arms_config_provider
    "arm_config": {
        "filename": None,
        "updated_at": None,
        # the arm ports associated with selected arm configuration
        "arm_parts": [],
    },
    # Structure of the objects in the array is consumed and controlled
    # by the UI.
    #
    # See hubState.js for details on the structure of the objects.
    "saved_positions": [],
}

persisted_state_keys: List[str] = [
    "arm_config_selected",
    "saved_positions",
]


# This should only be called by central_hub
def init_persisted_state() -> None:
    file = PERSISTED_STATE_FILE
    if c.STRONGARM_ENV == "test":
        log.info(
            f"Running in test mode, using {TEST_PERSISTED_STATE_INPUT_FILE} persisted state file"
        )
        file = TEST_PERSISTED_STATE_INPUT_FILE

    try:
        with open(file, "r") as f:
            persisted_state = json.load(f)
            for key in persisted_state_keys:
                if key in persisted_state:
                    state[key] = persisted_state[key]
    except FileNotFoundError:
        log.info("No persisted state file found")
    except Exception as e:
        log.error(f"Error loading persisted state: {e}")

    log.info(f"initial state: {state}")


# This should only be called by central_hub
def persist_state() -> None:
    """Persist specified state keys to file."""
    file = PERSISTED_STATE_FILE
    if c.STRONGARM_ENV == "test":
        log.info(
            f"Running in test mode, using {TEST_PERSISTED_STATE_OUTPUT_FILE} file to persist state"
        )
        file = TEST_PERSISTED_STATE_INPUT_FILE

    try:
        persisted_state = {
            key: state[key] for key in persisted_state_keys if key in state
        }
        with open(file, "w") as f:
            json.dump(persisted_state, f, indent=4)
    except (IOError, json.JSONDecodeError) as e:
        log.error(f"Failed to persist state: {e}")


def serialize_state(keysRequested=None) -> str:
    """Serialize current state to JSON string."""

    requestedState = state
    if keysRequested:
        requestedState = {key: state[key] for key in keysRequested if key in state}

    try:
        return json.dumps({"type": "state", "data": requestedState})
    except json.JSONDecodeError as e:
        log.error(f"Failed to serialize state: {e}")
        return json.dumps({"type": "state", "data": {}})


def update_state_from_message_data(message_data: Dict[str, Any]) -> None:
    global state
    """Update state from received message data."""
    for key, data in message_data.items():
        state[key] = data
        state[f"{key}_updated_at"] = time.time()

File: src/webapp_ts/src/util/hubState.ts
import "react";

const urlParams = new URLSearchParams(window.location.search);
const debugThings = urlParams.get("debug")?.split(",") || [];

export const logMessages = debugThings.indexOf("messages") >= 0;

// How often to check if hub is really still alive
const HUB_PING_INTERVAL = 1000;
// with us pinging every 1000ms, there should
// never be a lapse of more than 1500 between
// messages from hub.  Otherwise, we show "offline"
const MIN_HUB_UPDATE_INTERVAL = 1500;

export const HUB_PORT = process.env.HUB_PORT || 5800;

export const HUB_HOST = `${window.location.hostname}:${HUB_PORT}`;

export const HUB_URL = `ws://${HUB_HOST}/ws`;

export const DEFAULT_HUB_STATE = {
    // this is for the UI only
    hubConnStatus: "offline",

    // the keys below are shared from central hub.  See hub_state.py

    // provided by central_hub/
    hub_stats: { state_updates_recv: 0 },
    //last requested angles
    set_angles: [],
    // actual angles last reported by servos
    current_angles: [],

    // array of json file names provided to central hub
    // by arms_config_provider
    arm_config_files: [],

    // initially by arms_config_provider to last selected, can be changed
    // to any of the arm_configs
    arm_config_selected: "",

    // currently selected arm config provided by
    // arms_config_provider
    arm_config: {
        filename: "",
        description: "",
        arm_parts: [],
    },

    // saved positions provided initially by saved_positions_provider
    // and updated by the UI.   This should be an array of objects like:
    // {
    //   "name": "some name",
    //   "description": "some description",
    //   "angles": [0, 90, 90, 90, 90, 90]
    // }
    saved_positions: [],

    subsystem_stats: {},
};

let __hub_state = { ...DEFAULT_HUB_STATE };

let hubStatePromises: ((value: any) => void)[] = [];
let onUpdateCallbacks: ((state: typeof DEFAULT_HUB_STATE) => void)[] = [];
let lastHubUpdate = Date.now();
let hubMonitor: NodeJS.Timeout | null = null;

export let webSocket: WebSocket | null = null;

export function connectToHub(state = DEFAULT_HUB_STATE) {
    try {
        setHubConnStatus("connecting");
        console.log(`connecting to central-hub at ${HUB_URL}`);

        webSocket = new WebSocket(HUB_URL);

        webSocket.addEventListener("open", function (event) {
            lastHubUpdate = Date.now();

            try {
                webSocket!.send(JSON.stringify({ type: "getState" }));
                webSocket!.send(
                    JSON.stringify({ type: "identity", data: "webapp" })
                );
                webSocket!.send(
                    JSON.stringify({ type: "subscribeState", data: "*" })
                );
                setHubConnStatus("online");
            } catch (e) {
                onConnError(state, e);
            }
            startHubMonitor();
        });

        webSocket.addEventListener("error", function (event) {
            console.error("got error from central-hub socket", event);
        });

        webSocket.addEventListener("close", function (event) {
            onConnError(state, event);
        });

        webSocket.addEventListener("message", function (event) {
            lastHubUpdate = Date.now();
            let message = null;
            try {
                message = JSON.parse(event.data);
            } catch (e) {
                console.error("error parsing message from central-hub", e);
                return;
            }
            if (message.type === "pong") {
                return;
            }
            logMessage("got message from central-hub", {
                raw: event.data,
                parsed: message,
            });
            if (message.type === "state" && hubStatePromises.length > 0) {
                hubStatePromises.forEach((p) => p(message.data));
                hubStatePromises = [];
            } else if (
                message.type === "state" ||
                message.type === "stateUpdate"
            ) {
                updateStateFromCentralHub(message.data);
            }
        });
    } catch (e) {
        onConnError(state, e);
    }
}

function startHubMonitor() {
    stopHubMonitor();
    hubMonitor = setInterval(() => {
        // if the socket is hung or there is no network,
        // the websocket will not error out until we send something
        webSocket!.send(JSON.stringify({ type: "ping" }));

        if (
            __hub_state.hubConnStatus === "online" &&
            Date.now() - lastHubUpdate > MIN_HUB_UPDATE_INTERVAL
        ) {
            setHubConnStatus("offline");
        }
    }, HUB_PING_INTERVAL);
}

function stopHubMonitor() {
    if (hubMonitor) {
        clearInterval(hubMonitor);
    }
}

// handler gets called with __hub_state
export function addHubStateUpdatedListener(handler: (state: typeof DEFAULT_HUB_STATE) => void) {
    onUpdateCallbacks.push(handler);
}

export function removeHubStateUpdatedListener(handler: (state: typeof DEFAULT_HUB_STATE) => void) {
    const index = onUpdateCallbacks.indexOf(handler);
    if (index !== -1) {
        onUpdateCallbacks.splice(index, 1);
    }
}

export function getLocalState() {
    return __hub_state;
}

export function getStateFromCentralHub() {
    const statePromise = new Promise<typeof DEFAULT_HUB_STATE>((resolve) =>
        hubStatePromises.push(resolve)
    );
    webSocket!.send(JSON.stringify({ type: "getState" }));
    return statePromise;
}

export function updateSharedState(newState: Partial<typeof DEFAULT_HUB_STATE>) {
    webSocket!.send(JSON.stringify({ type: "updateState", data: newState }));
}

function delayedConnectToHub(state: typeof DEFAULT_HUB_STATE) {
    setTimeout(() => {
        if (state.hubConnStatus === "offline") {
            connectToHub(state);
        }
    }, 5000);
}

function onConnError(state: typeof DEFAULT_HUB_STATE, e: Event | Error) {
    console.error(
        "got close message from central-hub socket. will attempt to reconnnect in 5 seconds",
        e
    );
    stopHubMonitor();
    setHubConnStatus("offline");
    delayedConnectToHub(state);
}

// not exported, should only be called from connectToHub
function setHubConnStatus(newStatus: string) {
    logMessage("setting conn status", newStatus);
    __hub_state.hubConnStatus = newStatus;
    emitUpdated();
}

function updateStateFromCentralHub(hubData: Partial<typeof DEFAULT_HUB_STATE>) {
    for (const [key, value] of Object.entries(hubData)) {
        // TODO : maybe merge the state with incoming.
        // State for any top level key must be whole
        (__hub_state as any)[key] = value;
    }
    emitUpdated();
}

function emitUpdated() {
    for (const callback of onUpdateCallbacks) {
        callback(__hub_state);
    }
}

export function logMessage(...args: any[]) {
    if (logMessages) {
        console.log(...args);
    }
}
